opentelemetry-sdk/src/logs/log_processor.rs

```diff
@@ -167,6 +167,72 @@ enum BatchMessage {
 
 /// A [`LogProcessor`] that buffers log records and reports
 /// them at a pre-configured interval from a dedicated background thread.
+// **Memory Management in BatchLogProcessor**
+//
+// The `BatchLogProcessor` manages memory through the following stages of log processing:
+//
+// 1. **Record Ingestion**:
+//    - Each `LogRecord` is **cloned** upon entering the processor.
+//    - `LogRecordAttributes` utilize a hybrid memory model:
+//      - Attributes up to `PREALLOCATED_ATTRIBUTE_CAPACITY` are **stack-allocated**.
+//      - Exceeding attributes trigger **heap allocation** in a dynamically growing vector.
+//    - The `LogRecord` and its associated `InstrumentationScope` are **boxed together**
+//      to allocate them on the heap before entering the queue. Which means:
+//      - The entire `LogRecord` object, including its stack-allocated inline attributes, is moved to the heap.
+//      - Any overflow attributes already on the heap remain unaffected.
+//
+// 2. **Queue Management**:
+//    - Uses a **bounded synchronous channel** (`sync_channel`) with a maximum size defined by `max_queue_size`.
+//    - Each queued item is a **heap-allocated** `Box<(LogRecord, InstrumentationScope)>`.
+//    - When the queue is full:
+//      - Records are **dropped** (counted but not stored).
+//      - A warning is logged the first time this occurs.
+//
+// 3. **Worker Thread Storage**:
+//    - The worker thread maintains a pre-allocated `Vec` of boxed record pairs:
+//      - The vector’s capacity is fixed at `max_export_batch_size`.
+//    - Records are **moved** (not cloned) from the queue to the vector for processing.
+//
+// 4. **Export Process**:
+//    - A temporary `Vec` is created during the export process to hold ownership of the boxed records:
+//        - The contents of the worker thread's buffer (`Vec<Box<(LogRecord, InstrumentationScope)>>`)
+//          are moved into this temporary vector using `logs.split_off(0)`.
+//        - Ownership of the boxed records is transferred to the new vector, ensuring efficient
+//          memory usage without additional cloning.
+//    - The temporary vector is then used to construct references passed to the exporter via `LogBatch`.
```

Comment:
It'd be good to mention that there two temporary vectors created during each export:
1. One vector is created by `split_off(0)` (we could avoid this if we pursue #2483)
2. The second vector is created by collecting references from the first vector mentioned above.

I think we should be able to avoid the second vector creation as well by updating `LogBatch` constructors.